README.txt
Pinder 0.6.5
============

Pinder is a client library for Campfire, the chat application from 37Signals.

You can find Campfire here: <http://www.campfirenow.com/>

Requirements
------------

* Python 2.3+
* BeautifulSoup >= 3.0.4
* httplib2 >= 0.3.0

Installation
------------


$ easy_install -U BeautifulSoup
$ easy_install -U httplib2

Then:

$ easy_install -U pinder

- OR -

Get the latest version from <http://dev.oluyede.org/download/pinder/>, unpack it, go inside the new directory and type the following in the shell:

$ python setup.py install


Documentation
-------------

See doc/usage.txt for usage documentation and doc/api/pinder.html for the API doc. There's also a faq in doc/faq.txt.


Many thanks to Brandon Keepers <http://opensoul.org/> for Tinder.

Subversion access
-----------------

See <http://dev.oluyede.org/trac/default/wiki/Pinder>


NEWS.txt
Pinder 0.6.5 [May 18 2007]:
    - BACKWARDS INCOMPATIBLE CHANGE: BeautifulSoup and httplib2 are no longer
      distributed with Pinder. You have to install them separately.

    - Fixed a bug in Room.leave()
    
    - Fixed a couple of bugs in the tests
    
    - Room objects now have ping() to ping the server and topic() to retrieve
      the topic of the room

Pinder 0.6.0 [Apr 08 2007]:
    - Campfire objects now have rooms() and rooms_names() methods to get the 
      list of the associated room objects and the names of all the rooms
    
    - Campfire objects also have find_or_create_room_by_name(), an helper 
      method which combine find_room_by_name() and create_room()
     
    - The whole library has been updated to httlibp2 0.3.0
    
    - A proper user agent is sent during the requests
    
    - Room objects now have guest_access_enabled() to know if the guests can
      enter that room
    
    - The support for transcripts has been added throughout the library. See
      the changelog for details.


test/utils.py
import os
from runtests import TESTDIR

FIXTURE = 'default'

class MockResponse(object):
    def __init__(self):
        self.status = 200
        self.headers = {
            'location': '/foobar',
            'set-cookie': 'cookie',
        }
        self.fixture = ''
    
    def getheader(self, header_name, default=None):
        return self.headers.get(header_name)
    get = getheader

    def read(self):
        path = os.path.join(TESTDIR, "fixtures/%s.html" % FIXTURE)
        return open(path).read()

class MockHttplib2Response(MockResponse):
    def __init__(self, info):
        self.status = info.status
        self.headers = info.headers
        self.fixture = ''
    
    def has_key(self, key):
        return self.__contains__(key)
        
    def __iter__(self):
        return iter(self.headers.keys())
        
    def __contains__(self, item):
        return item in self.headers
        
    def __getitem__(self, item):
        item = self.headers.get(item)
        if not item:
            raise KeyError(item)
        
    def __setitem__(self, key, value):
        self.headers[key] = value


test/runtests.py
#! /usr/bin/env python
import os
import sys
import unittest

if __name__ == "__main__":
    __file__ = sys.argv[0]

TESTDIR = os.path.dirname(os.path.abspath(__file__))
TOPDIR = os.path.dirname(TESTDIR)

def load_test_files():
    test_files = []
    for f in os.listdir(TESTDIR):
        if f.startswith('test_') and f.endswith('.py'):
            test_files.append(os.path.splitext(f)[0])
    return test_files
    
if __name__ == "__main__":
    if TOPDIR not in sys.path:
        sys.path.insert(0, TOPDIR)
    sys.path.insert(0, TESTDIR)
    suite = unittest.defaultTestLoader.loadTestsFromNames(load_test_files())
    try:
        v = sys.argv[1]
    except:
        v = 1
    runner = unittest.TextTestRunner(verbosity=int(v))
    runner.run(suite)


test/test_room.py
from httplib import HTTPConnection
import unittest

from pinder import Campfire, Room
import utils


class RoomTest(unittest.TestCase):
    def setUp(self):
        self.response = utils.MockResponse()
        self.campfire = Campfire('foobar')
        self.room = Room(self.campfire, 12345, 'Room 1')
        self.campfire = Campfire('foobar')
        response = self.response
        HTTPConnection.request = lambda self, m, l, b, h: None
        HTTPConnection.getresponse = lambda self: response
        
    def test_join(self):
        utils.FIXTURE = 'room_info'
        self.assertEqual(True, self.room.join())
        self.assertEqual(True, self.room.join(force=True))

    def test_guest_url_no_guest_url(self):
        utils.FIXTURE = 'no_guest_url'
        self.assert_(self.room.guest_url() is None)
        
    def test_guest_url(self):
        utils.FIXTURE = 'guest_url'
        self.assertEqual('http://sample.campfirenow.com/99d14',
            self.room.guest_url())
            
    def test_guest_invite_code_no_guest_url(self):
        utils.FIXTURE = 'no_guest_url'
        self.assert_(self.room.guest_invite_code() is None)
        
    def test_guest_invite_code(self):
        utils.FIXTURE = 'guest_url'
        self.assertEqual('99d14', self.room.guest_invite_code())
        
    def test_transcripts(self):
        utils.FIXTURE = 'transcripts'
        transcripts = self.room.transcripts()
        self.assertEqual(1, len(transcripts))
        

if __name__ == '__main__':
    unittest.main()


pinder/room.py
"""
Handles the Campfire room.
"""
from datetime import datetime
import re
import time
import urlparse

from BeautifulSoup import BeautifulSoup

class Room(object):
    "A Campfire room."
    def __init__(self, campfire, id, name=None):
        self._campfire = campfire
        self._verify_response = campfire._verify_response
        self._post = campfire._post
        self._get = campfire._get
        self._room = None

        self.membership_key = self.user_id = None
        self.last_cache_id = self.timestamp = None
        self.idle_since = datetime.now()

        #: The room id.
        self.id = id
        #: The name of the room.
        self.name = name
        #: The U{urlparsed<http://docs.python.org/lib/module-urlparse.html#l2h-4268>} URI of the room.
        self.uri = urlparse.urlparse("%s/room/%s" % (urlparse.urlunparse(self._campfire.uri), self.id))

    def __repr__(self):
        return "<Room: %s>" % self.name

    def __eq__(self, other):
        return self.id == other.id

    def join(self, force=False):
        """Joins the room; if 'force' is True join even if already joined.

        Returns True if successfully joined, False otherwise. """
        if not self._room or force:
            self._room = self._get("room/%s" % self.id)
            if not self._verify_response(self._room, success=True):
                self._room = None
                return False
            self._get_room_data()
        self.ping()
        return True

    def leave(self):
        """Leaves the room.

        Returns True if successfully left, False otherwise."""
        has_left = self._verify_response(
            self._post('room/%s/leave' % self.id), redirect=True)
        self._room = self.membership_key = self.user_id = None
        self.last_cache_id = self.timestamp = self.idle_since = None
        return has_left

    def toggle_guest_access(self):
        """Toggles guest access on and off.

        Returns True if successfully toggled, False otherwise."""
        response_result = self._verify_response(
            self._post(
                'room/%s/toggle_guest_access' % self.id), success=True)

        # go back inside to toggle the guest access
        return response_result and self.join(True)

    def guest_access_enabled(self):
        """Checks if the guest access is enabled.

        Returns True if the guest access is enabled, False otherwise."""
        return self.guest_url() is not None

    def guest_url(self):
        """Gets the URL for guest access.

        Returns None if the guest access is not enabled."""
        self.join()
        soup = BeautifulSoup(self._room.body)
        try:
            return soup.find('div', {'id': 'guest_access_control'}).h4.string
        except AttributeError:
            return None

    def guest_invite_code(self):
        """Gets the invite code for guest access.

        Returns None if the guest access is not enabled."""
        url = self.guest_url()
        if url:
            return re.search(r'\/(\w*)$', url).groups(0)[0]

    def change_name(self, name):
        """Changes the name of the room.

        Returns the new name if successfully changed, None otherwise."""
        response = self._post('account/edit/room/%s' % self.id,
            {'room[name]': name}, ajax=True)
        if self._verify_response(response, success=True):
            self.name = name
            return self.name
    rename = change_name

    def change_topic(self, topic):
        """Changes the topic of the room.

        Returns the new topic if successfully changed, None otherwise."""
        response = self._post('room/%s/change_topic' % self.id,
            {'room[topic]': topic}, ajax=True)
        if self._verify_response(response, success=True):
            return topic

    def topic(self):
        """Gets the current topic, if any."""
        self.join()
        soup = BeautifulSoup(self._room.body)
        h = soup.find(attrs={'id': 'topic'})
        if h:
            def _is_navigable_string(tag):
                return not hasattr(tag, 'attrs')
            topic = filter(_is_navigable_string, h.contents)
            return repr("".join(topic).strip())

    def lock(self):
        """Locks the room to prevent new users from entering.

        Returns True if successfully locked, False otherwise."""
        return self._verify_response(self._post(
            'room/%s/lock' % self.id, {}, ajax=True), success=True)

    def unlock(self):
        """Unlocks the room.

        Returns True if successfully unlocked, False otherwise."""
        return self._verify_response(self._post(
            'room/%s/unlock' % self.id, {}, ajax=True), success=True)

    def ping(self, force=False):
        """Pings the server updating the last time we have been seen there.

        Returns True if successfully pinged, False otherwise."""
        now = datetime.now()
        delta = now - self.idle_since
        if delta.seconds < 60 or force:
            self.idle_since = datetime.now()
            return self._verify_response(self._post(
                'room/%s/tabs' % self.id, {}, ajax=True), success=True)
        return False

    def destroy(self):
        """Destroys the room.

        Returns True if successfully destroyed, False otherwise."""
        return self._verify_response(self._post(
            'account/delete/room/%s' % self.id), success=True)

    def users(self):
        """Lists the users chatting in the room.

        Returns a set of the users."""
        return self._campfire.users(self.name)

    def speak(self, message):
        """Send a message to the room.

        Returns the message if successfully sent it, None otherwise."""
        self.join()
        return self._send(message)

    def paste(self, message):
        """Paste a message to the room.

        Returns the message if successfully pasted it, None otherwise."""
        self.join()
        return self._send(message, {'paste': True})

    def messages(self):
        """Gets new messages.

        Returns a list of message data:
         * id: the id of the message
         * person: the name of the person who wrote the message if any
         * user_id: the user id of the person if any
         * message: the message itself if any"""
        data = dict(l=self.last_cache_id, m=self.membership_key,
                    s=self.timestamp, t=int(time.time()))
        response = self._post("poll.fcgi", data, ajax=True)

        cache_match = re.search(r'lastCacheID = (\d+)', response.body)
        if cache_match:
            self.last_cache_id = cache_match.groups(0)[0]

        messages = []

        for line in response.body.split("\r\n"):
            if 'timestamp_message' in line:
                continue

            id_match = re.search(r'message_(\d+)', line)
            if not id_match:
                continue

            try:
                messages.append(dict(
                    id = id_match.groups(0)[0],
                    user_id = re.search(r'user_(\d+)', line).groups(0)[0],
                    person = re.search(r'\\u003Ctd class=\\"person\\"\\u003E(?:\\u003Cspan\\u003E)?(.+?)(?:\\u003C\/span\\u003E)?\\u003C\/td\\u003E', line).groups(0)[0],
                    message = re.search(r'\\u003Ctd class=\\"body\\"\\u003E\\u003Cdiv\\u003E(.+?)\\u003C\/div\\u003E\\u003C\/td\\u003E', line).groups(0)[0]
                    ))
            except AttributeError:
                continue

        return messages

    def transcripts(self):
        """Gets the dates of transcripts of the room.

        Returns a list of dates."""
        return self._campfire.transcripts(self.id)

    def transcript(self, date):
        """Get the transcript for the given date (a datetime.date instance).

        Returns a list of message data:
         * id: the id of the message
         * person: the name of the person who wrote the message if any
         * user_id: the user id of the person if any
         * message: the message itself if any"""
        uri = 'room/%s/transcript/%s' % (self.id, date.strftime('%Y/%m/%d'))

        soup = BeautifulSoup(self._get(uri).body)

        def _filter_messages(tag):
            return tag.has_key('class') and 'message' in tag['class'].split()
        messages = soup.findAll(_filter_messages)

        all_transcript = []
        for message in messages:
            t = {}

            person = message.find(True, attrs={'class': 'person'})
            try:
                t['person'] = person.span.string
            except AttributeError:
                try:
                    t['person'] = person.string
                except AttributeError:
                    t['person'] = None

            body = message.find('td', attrs={'class': 'body'})
            try:
                t['message'] = body.div.string
            except AttributeError:
                t['message'] = None

            t['id'] = re.search(r'message_(\d+)', message['id']).groups()[0]
            match = re.search(r'user_(\d+)', message['class'])
            if match:
                t['user_id'] = match.groups()[0]
            else:
                t['user_id'] = None

            all_transcript.append(t)

        return all_transcript

    def _send(self, message, options={}):
        data = {'message': message, 't': int(time.time())}
        data.update(options)
        response = self._post('room/%s/speak' % self.id, data, ajax=True)
        if self._verify_response(response, success=True):
            return message

    def _get_room_data(self):
        self.membership_key = re.search(r'\"membershipKey\": \"([a-z0-9]+)\"',
            self._room.body).groups(0)[0]
        self.user_id = re.search(r'\"userID\": (\d+)',
            self._room.body).groups(0)[0]
        self.last_cache_id = re.search(r'\"lastCacheID\": (\d+)',
            self._room.body).groups(0)[0]
        self.timestamp = re.search(r'\"timestamp\": (\d+)',
            self._room.body).groups(0)[0]



__all__ = ['Room']


pinder/campfire.py
"""
Handles Campfire online chat.
"""
import datetime
import re
import time
import urllib
import urlparse

from BeautifulSoup import BeautifulSoup
import httplib2

try:
    set # python 2.3 does not have the set data type
except NameError:
    from sets import Set as set

from __init__ import __version__
from room import Room

class Campfire(object):
    "Creates a connection to the Campfire account with the given subdomain."
    def __init__(self, subdomain):
        #: The Campfire's subdomain.
        self.subdomain = subdomain
        #: True if the user is logged in Campfire, False otherwise.
        self.logged_in = False
        #: Contains the connection cookie.
        self.cookie = None
        #: The U{urlparsed<http://docs.python.org/lib/module-urlparse.html#l2h-4268>} URI of the Campfire account.
        self.uri = urlparse.urlparse("http://%s.campfirenow.com" % self.subdomain)
        self._location = None
        self._room_re = re.compile(r'room\/(\d*)')
        self._http_client = httplib2.Http(timeout=5)
        self._http_client.force_exception_to_status_code = False
    
    def login(self, email, password):
        """Logs into Campfire with the given email and password.
        
        Returns True if logged in, False otherwise."""
        response = self._post("login",
            dict(email_address=email, password=password))
        self.logged_in = self._verify_response(response,
            redirect_to=self._uri_for())
        return self.logged_in
            
    def logout(self):
        """Logs out from Campfire.
        
        Returns True if logged out, False otherwise."""
        retval = self._verify_response(self._get("logout"), redirect=True)
        self.logged_in = not retval
        return retval
    
    def create_room(self, name, topic=''):
        """Creates a Campfire room with the given name and an optional topic.
        
        Returns None if the room has not been created."""
        room_data = {
            'room[name]': name,
            'room[topic]': topic
        }
        response = self._post('account/create/room?from=lobby', room_data,
            ajax=True)
        if self._verify_response(response, success=True):
            return self.find_room_by_name(name)
            
    def find_room_by_name(self, name):
        """Finds a Campfire room with the given name.
        
        Returns a Room instance if found, None otherwise."""
        rooms = self._get_rooms_markup()
        
        for room in rooms:
            try:
                room_name = room.h2.a.string
            except AttributeError: # the chat is full
                room_name = room.h2.string.strip()

            if room_name.lower() == name.lower():
                try:
                    room_uri = room.h2.a['href']
                except AttributeError: # no uri available (!?)
                    return None
                
                room_id = self._room_id_from_uri(room_uri)
                return Room(self, room_id, name)
                
    def find_or_create_room_by_name(self, name):
        """Finds a Campfire room with the given name.
        If the room is not present it will be created.
        
        Returns a Room instance."""
        return self.find_room_by_name(name) or self.create_room(name)
    
    def users(self, *room_names):
        """Lists the users chatting in any room or in the given room(s).
        
        Returns a set of the users."""
        rooms = self._get_rooms_markup()
         
        all_users = []
        for room in rooms:
            try:
                room_name = room.h2.a.string
            except AttributeError: # the chat is full
                room_name = room.h2.string.strip()
            if not room_names or room_name in room_names:
                room_users = room.find('ul').findAll('span')
                for user in room_users:
                    all_users.append(user.string)
        return set(all_users)

    def rooms_names(self):
        """Lists the names of the rooms available in the Campfire subdomain.

        Returns a list of the names of the rooms."""
        rooms = self._get_rooms_markup()

        rooms_names_list = []
        for room in rooms:
            try:
                rooms_names_list.append(room.h2.a.string)
            except AttributeError:
                rooms_names_list.append(room.h2.string.strip())
        rooms_names_list.sort()
        return rooms_names_list
        
    def rooms(self):
        """Lists the available rooms.
        
        Returns a list of Room objects."""
        names = self.rooms_names()
        return [self.find_room_by_name(name) for name in names]

    def transcripts(self, room_id=None):
        """Gets the dates of the transcripts by room filtered by the given id
        if any.
        
        Returns a dictionary of all the dates by room.
        """
        uri = 'files%2Btranscripts'
        if room_id:
            uri = '%s?room_id=%s' % (uri, str(room_id))

        soup = BeautifulSoup(self._get(uri).body)
        
        def _filter_transcripts(tag):
            return tag.has_key('class') and 'transcript' in tag['class'].split()
        transcripts_markup = soup.findAll(_filter_transcripts)

        result = {}
        for tag in transcripts_markup:
            link = tag.a['href']
            found_room_id = self._room_id_from_uri(link)
            date = re.search(
                r'/transcript/(\d{4}/\d{2}/\d{2})', link).groups()[0]
            try:
                result[found_room_id]
            except KeyError:
                result[found_room_id] = []
            result[found_room_id].append(self._parse_transcript_date(date))
        
        if room_id:
            return result[str(room_id)]
        return result
            
    def _parse_transcript_date(self, date):
        return datetime.date.fromtimestamp(time.mktime(
            time.strptime(date, '%Y/%m/%d')))

    def _filter_rooms_markup(self, tag):
        return tag.name == 'div' and tag.has_key('id') and tag['id'].startswith('room_')
            
    def _get_rooms_markup(self):
        body = self._get().body
        soup = BeautifulSoup(body)
        return soup.findAll(self._filter_rooms_markup)
    
    def _uri_for(self, path=''):
        return "%s/%s" % (urlparse.urlunparse(self.uri), path)
         
    def _room_id_from_uri(self, uri):
        try:
            return self._room_re.split(uri)[1]
        except IndexError:
            pass
        
    def _prepare_request(self, **options):
        headers = {}
        
        headers['User-Agent'] = 'Pinder/%s' % __version__

        if self.cookie:
            headers['cookie'] = self.cookie
        if 'ajax' in options:
            headers['X-Requested-With'] = 'XMLHttpRequest'
            headers['X-Prototype-Version'] = '1.5.1.1'
        return headers
        
    def _perform_request(self, method, path, data={}, **options):
        headers = self._prepare_request(**options)
        if method == 'POST':
            headers.update({"Content-type": "application/x-www-form-urlencoded"})
            location = self._uri_for(path)
        elif method == 'GET':
            if self._location:
                location = self._location
            else:
                location = self._uri_for(path)
            self._location = None
        else:
            raise Exception, 'Unsupported HTTP method'

        response, content = self._http_client.request(location, method,
            urllib.urlencode(data), headers)
        response.body = content

        if response.get('set-cookie'):
            self.cookie = response.get('set-cookie')

        return response

    def _post(self, path, data={}, **options):
        return self._perform_request('POST', path, data, **options)

    def _get(self, path=''):
        return self._perform_request('GET', path)

    def _verify_response(self, response, **options):
        if 'success' in options:
            return response.status == 200
        elif 'redirect' in options:
            return response.status in xrange(300, 400)
        elif 'redirect_to' in options:
            location = response.get('location')
            return self._verify_response(
                response, redirect=True) and location == options['redirect_to']
        else:
            return False
            

__all__ = ['Campfire']
